File: SOLID/dependency_inversion.cpp
#include<iostream>
using namespace std;

///////////// Without Dependency Inversion Principle ///////////////
namespace WithoutDependencyInversion {
    class LowLevel {
    public:
        void doSomething() {
            cout << "LowLevel::doSomething()" << endl;
        }
    };

    /*
        This class directly uses the low level class.
        We cannot use any other class in place of LowLevel without
        making changes in the definition.
    */
    class HighLevel {
    public:
        void UseLowLevel(LowLevel lowLevel) {
            lowLevel.doSomething();
        }
    };
}

//////////// With Dependency Inversion //////////////////
namespace WithDependencyInversion {
    class LowLevelBase {
    public:
        virtual void doSomething() = 0;
    };

    class LowLevel1: public LowLevelBase {
    public:
        void doSomething() {
            cout << "LowLevel1::doSomething()" << endl;
        }
    };

    class LowLevel2: public LowLevelBase {
    public:
        void doSomething() {
            cout << "LowLevel2::doSomething()" << endl;
        }
    };

    /*
        This class uses the base class, so we can use any class
        derived from the base class in place of LowLevel.
    */
    class HighLevel {
    public:
        void UseLowLevel(LowLevelBase* lowLevel) {
            lowLevel->doSomething();
        }
    };
}

int main() {
    cout << "Without Dependency Inversion Principle" << endl;
    WithoutDependencyInversion::HighLevel highLevel;
    highLevel.UseLowLevel(WithoutDependencyInversion::LowLevel());

    cout << "With Dependency Inversion Principle" << endl;
    WithDependencyInversion::HighLevel highLevel2;
    highLevel2.UseLowLevel(new WithDependencyInversion::LowLevel1());
    highLevel2.UseLowLevel(new WithDependencyInversion::LowLevel2());

    return 0;
}

File: SOLID/dependency_inversion.cs
using System;
using System.Console;


//////////////////////////////////////////////////// WITHOUT DEPENDENCY INVERSION //////////////////////////////////////
namespace WithoutDependencyInversion {
    
    public class LowLevel {
        public void ImportantWork() {
            System.Console.WriteLine("Important work done by low level component...");
        }
    }

    public class HighLevel {
        public static void UseLowLevel(LowLevel lowLevelObject) {
            lowLevelObject.ImportantWork();
        }
    }


    public static class Program {
        static void Main(string[] args) {
            Console.WriteLine(HighLevel.UseLowLevel(new LowLevelImplementation1()));
        }
    }
}

//////////////////////////////////////////////////// WITH DEPENDENCY INVERSION //////////////////////////////////////
namespace WithDependencyInversion {
    
    // Abstraction for Low level component
    public interface ILowLevel {
        public void ImportantWork();
    }

    public class LowLevelImplementation1: ILowLevel {
        // This implements an abstract class instead of being a 
        // direct concrete class
        public void ImportantWork() {
            System.Console.WriteLine("Important work done by low level component 1...");
        }
    }
    
    public class LowLevelImplementation2: ILowLevel {
        // This implements an abstract class instead of being a 
        // direct concrete class
        public void ImportantWork() {
            Console.WriteLine("Important work done by low level component 2...");
        }
    }

    public static class HighLevel {
        // The High level class now is dependent on an abstract rather
        // than a concrete implementation
        public static void UseLowLevel(ILowLevel lowLevelObject) {
            lowLevelObject.ImportantWork();
        }
    }


    public static class Program {
        static void Main(string[] args) {
            // Since abstraction is used now, a different implementation
            // of low level class can be used with the high level component
            Console.WriteLine(HighLevel.UseLowLevel(new LowLevelImplementation1()));
            Console.WriteLine(HighLevel.UseLowLevel(new LowLevelImplementation2()));
        }
    }
}